pkg/controller/vitessbackupschedule/vitessbackupschedule_controller.go

```diff
@@ -0,0 +1,624 @@
+}
+
+func createVtctldClientCommand(cmd *strings.Builder, serverAddr string, extraFlags map[string]string, keyspace, shard string) {
+	cmd.WriteString(fmt.Sprintf("%s %s BackupShard", vtctldclientPath, serverAddr))
+
+	// Add any flags
+	for key, value := range extraFlags {
+		cmd.WriteString(fmt.Sprintf(" --%s=%s", key, value))
+	}
+
+	// Add keyspace/shard or tablet alias
+	cmd.WriteString(fmt.Sprintf(" %s/%s", keyspace, shard))
+}
+
+func (r *ReconcileVitessBackupsSchedule) getVtctldServiceName(ctx context.Context, vbsc *planetscalev2.VitessBackupSchedule, cluster string) (svcName string, svcPort int32, err error) {
+	svcList := &corev1.ServiceList{}
+	listOpts := &client.ListOptions{
+		Namespace: vbsc.Namespace,
+		LabelSelector: apilabels.Set{
+			planetscalev2.ClusterLabel:   cluster,
+			planetscalev2.ComponentLabel: planetscalev2.VtctldComponentName,
+		}.AsSelector(),
+	}
+	if err = r.client.List(ctx, svcList, listOpts); err != nil {
+		return "", 0, fmt.Errorf("unable to list vtctld service in %q: %v", vbsc.Namespace, err)
+	}
+
+	if len(svcList.Items) > 0 {
+		service := svcList.Items[0]
+		svcName = service.Name
+		for _, port := range service.Spec.Ports {
+			if port.Name == planetscalev2.DefaultGrpcPortName {
+				svcPort = port.Port
+				break
+			}
+		}
+	}
+
+	if svcName == "" || svcPort == 0 {
+		return "", 0, fmt.Errorf("no vtctld service found in %q namespace", vbsc.Namespace)
+	}
+	return svcName, svcPort, nil
+}
+
+func (r *ReconcileVitessBackupsSchedule) getAllShardsInKeyspace(ctx context.Context, namespace, cluster, keyspace string) ([]string, error) {
+	shardsList := &planetscalev2.VitessShardList{}
+	listOpts := &client.ListOptions{
+		Namespace: namespace,
+		LabelSelector: apilabels.Set{
+			planetscalev2.ClusterLabel:  cluster,
+			planetscalev2.KeyspaceLabel: keyspace,
+		}.AsSelector(),
+	}
+	if err := r.client.List(ctx, shardsList, listOpts); err != nil {
+		return nil, fmt.Errorf("unable to list shards of keyspace %s in %s: %v", keyspace, namespace, err)
+	}
+	var result []string
+	for _, item := range shardsList.Items {
+		result = append(result, item.Spec.Name)
+	}
+	return result, nil
+}
+
+type keyspace struct {
+	name   string
+	shards []string
+}
+
+func (r *ReconcileVitessBackupsSchedule) getAllShardsInCluster(ctx context.Context, namespace, cluster string) ([]keyspace, error) {
+	ksList := &planetscalev2.VitessKeyspaceList{}
+	listOpts := &client.ListOptions{
+		Namespace: namespace,
+		LabelSelector: apilabels.Set{
+			planetscalev2.ClusterLabel: cluster,
+		}.AsSelector(),
+	}
+	if err := r.client.List(ctx, ksList, listOpts); err != nil {
+		return nil, fmt.Errorf("unable to list keyspaces in namespace %s: %v", namespace, err)
+	}
+	var result []keyspace
+	for _, item := range ksList.Items {
+		ks := keyspace{
+			name: item.Spec.Name,
+		}
+		for shardName := range item.Status.Shards {
+			ks.shards = append(ks.shards, shardName)
+		}
+		if len(ks.shards) > 0 {
+			result = append(result, ks)
+		}
```

Comment:
Curious why we don't do this instead:
```
		for shardName := range item.Status.Shards {
			ks.shards = append(result, &keyspace{
			name: item.Spec.Name,
			shards: shardName,
		})
		}
```

The other allocations/copying seems unnecessary at first glance. When combined with the single shot precise allocation it should be more efficient.